development/extensions/tutorial_key_concepts.rst

```diff
@@ -0,0 +1,316 @@
+
+This tutorial explains:
+
+ * `Dependency injection`_
+ * `PHP files`_
+ * `Template files`_
+ * `Language files`_
+ * `Javascript and CSS files`_
+
+
+Dependency injection
+====================
+
+A large portion of the phpBB codebase since 3.1 is based on a fundamental mechanism, namely dependency injection. The
+goal is to not have to worry about managing the different dependencies of the components you want to use (e.g.: the
+database layer, the cache system or the current user object). It's also a means to cease using globally scoped
+variables.
+
+Since phpBB 3.1 this has been implemented using the Symfony DependencyInjection Component. All services provided by
+phpBB are defined in a YAML format file in ``/config/services.yml``. Extensions can use the services defined by phpBB as
+well as define their own services for their PHP classes in their own ``config/services.yml``.
+
+To provide a simple example of dependency injection in action, let's look at the constructor method for Acme Demo's
+controller class ``main.php``:
+
+.. code-block:: php
+
+    public function __construct(\phpbb\config\config $config, \phpbb\controller\helper $helper, \phpbb\template\template $template, \phpbb\user $user)
+    {
+        $this->config = $config;
+        $this->helper = $helper;
+        $this->template = $template;
+        $this->user = $user;
+    }
+
+As you can see, it requires the ``$config``, ``$helper``, ``$template``, and ``$user`` objects
+(aka services). To have these services automatically resolved by dependency injection, The Acme Demo extension must
+define its controller class as a service in the following manner:
+
+.. code-block:: yaml
+
+    services:
+        acme.demo.controller:
+            class: acme\demo\controller\main
+            arguments:
+                - '@config'
+                - '@controller.helper'
+                - '@template'
+                - '@user'
+
+Note that ``acme.demo.controller`` is the unique service name for the class. Service names
+can be whatever you want, but phpBB extensions should prefix them with the vendor and extension name (in this case,
+``acme.demo``) to prevent potential conflicts with other extensions or core services.
+
+The ``class`` is simply the name-spaced path to the class.
+
+The ``arguments`` are the service dependencies required by the class. It is important that the
+order of arguments must match the order of parameters in the ``__construct()`` method definition.
+It is also recommended that any arguments beginning with ``@`` or ``%`` symbols be wrapped in single quotes.
+
+Most PHP files in an extension should use the dependency injection model with service definitions, particularly
+controllers and event listeners. The exceptions to this are any ACP/MCP/UCP files, language files and migration files
+(which do not currently use the service container).
+
+.. seealso::
+
+    * `Dependency Injection Container Wiki <https://wiki.phpbb.com/Dependency_Injection_Container>`_
```

Comment:
I would rather have the links to the pages, and change them later. Rather than not have them at all and forget to ever add them back IF they ever get removed from the wiki ;)